### PR TITLE
[refactor]: Upgrade to support v18.X of community EKS module

### PR DESCRIPTION
**Pull request by jrsdav**
Fixes:
- https://github.com/streamnative/terraform-aws-cloud/issues/53
- https://github.com/streamnative/terraform-aws-cloud/issues/84 

## Motivation
This Terraform module relies on a community driven AWS EKS module and has been steadily using version `v17.X` for the majority of our releases.

In the most recent version of the community EKS module, `v18.X`, [significant breaking changes](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/UPGRADE-18.0.md) were introduced which prevented our ability to continue staying current with updates.

It's been a long desire to re-work this module to be compatible with `v18.X` of the community EKS module, which also presented a good opportunity for us to refine our configuration approach for StreamNative platform.

This PR brings in the necessary changes to make this module compatible with `v18.X`, along with significant (and opinionated) changes to refine our use and configuration of AWS EKS environments compatible with StreamNative.

### Default Behavior
The behavior of the module has changed significantly. Using a "Vanilla" configuration, the following will occur:

#### All Resources
- The input `additional_tags` applies custom sets of AWS resource tags against _all applicable_ resources created by this module.

#### Cluster addons (`autoscaler`, `cert-manager`, `csi`, `external-dns`, `metrics-server`, `node-termination-handler`):
- The cluster addons will continue to be installed by default
- Addons will create an IAM role & IAM Policy

#### Node Groups
- Desired node group size is set to `0` (with the exception of the `small` node group, which is set to 1).
- Root volume disk size now defaults to `100GiB`
- Node groups will be created for every availability zone _and_ every instance class defined by the `node_pool_instance_types` default value (e.g. `3 AZs` * `4 Instance Classes` = `12 node groups`)

**Note:** The node groups changed in this module will result in Terraform wanting to _destroy_ existing node groups. Because of this, it is recommended that you _remove_ node group references in existing configurations before upgrading to this version (you will have to manually migrate away from and decommission the legacy node groups). An upgrade guide will be provided with more details.

### Modified Behavior
Here is a brief overview of what is now able to be modified as part of the module:

- The (now legacy) `use_runtime_mode` input is sticking around for backwards compatibility. It's sole responsibility is for management of the EKS cluster IAM role in _this_ module, as opposed to being created/managed in the community EKS module. Continue using it as `true` if done so previously. 
- The `enable_bootstrap` input controls _all_ of the add-ons installed by this module, including Istio (unless overriden by specifying `enable_istio = true`). If `enable_bootstrap` is set to `false`, nothing will be installed on the cluster (this allows us to use ArgoCD for installing add-ons for StreamNative's own configurations)
- When the `create_iam_policies` input is set to `false`, the module will _attach_ an IAM policy to an add-on IRSA IAM role, rather than try to create one. It defaults to IAM policies named `StreamNativeCloudRuntimePolicy` and `StreamNativeCloudLBPolicy`, but can also have a custom policy ARN provided using the `sncloud_services_iam_policy_arn` and `sncloud_services_lb_policy_arn` inputs.

### Removed
The following has been _removed_ from the module:

#### Resources:
- Calico
- External Secrets
- Functions Node Group
- Custom CloudWatch Configurations

#### Sub-modules:
  - `backup-resources` (added directly in the main module)
  - `managed-cloud` (moved to github.com/streamnative/terraform-managed-cloud)
  - `vault-resources` (no longer used or needed)
  - `tiered-storage-resources` (added directly in the main module

#### Variables:
- `aws_partition`
- `calico_*` (multiple inputs)
- `cluster_log_kms_key_id`
- `disk_encryption_kms_key_id` (renamed to `disk_encryption_kms_key_arn`)
- `enable_aws_load_balancer_controller`
- `enable_calico` 
- `enable_cert_manager` (now installed with `enable_bootstrap`)
- `enable_cluster_autoscaler`  (now installed with `enable_bootstrap`)
- `enable_csi`  (now installed with `enable_bootstrap`)
- `enable_cert_manager`  (now installed with `enable_bootstrap`)
- `enable_cluster_autoscaler`  (now installed with `enable_bootstrap`)
- `enable_external_secrets` 
- `enable_external_dns`  (now installed with `enable_bootstrap`)
- `enable_func_pool_monitoring`
- `enable_metrics_server`  (now installed with `enable_bootstrap`)
- `external_secrets_*` (multiple values)
- `extra_node_pool_instance_types`
- `func_pool_*` (multiple values)
- `istio_network_loadbancer ` (now controlled with `disable_public_pulsar_endpoint`)
- `map_additional_aws_accounts`
- `map_additional_iam_users`
- `node_pool_ami_is_eks_optimized` (renamed to `node_pool_ebs_optimized`)
- `wait_for_cluster_timeout`

#### Outputs
- `cloudwatch_log_group_arn`
- `node_groups` (changed to `eks_node_groups`)
- `worker_iam_role_arn`
- `worker_security_group_id` (changed to `eks_node_group_security_group_id`)
- `worker_https_ingress_security_group_rule`

### Added
The following has been _added_ to the module:

Resources:
- Tiered Storage (used Pulsar topic offloading)
- Velero (used for backups)

#### Variables:
- `s3_encryption_kms_key_arn`
- `cluster_security_group_additional_rules`
- `cluster_security_group_id`
- `create_cluster_security_group`
- `create_iam_policies`
- `create_node_security_group`
- `disable_public_pulsar_endpoint`
- `disable_public_eks_endpoint`
- `disk_encryption_kms_key_arn`
- `enable_bootstrap`
- `enable_sncloud_control_plane_access`
- `node_security_group_additional_rules`
- `node_security_group_id` 
- `node_pool_disk_iops`
- `node_pool_ebs_optimized`
- `velero_backup_schedule`
- `velero_excluded_namespaces`
- `velero_helm_chart_name`
- `velero_helm_chart_repository`
- `velero_helm_chart_version`
- `velero_namespace`
- `velero_plugin_version`
- `velero_policy_arn`
- `velero_settings`

#### Outputs:
- `eks_cluster_endpoint`
- `eks_cluster_platform_version`
- `eks_node_group_iam_role_arn`
- `eks_node_group_security_group_id`
- `eks_node_groups`
- `tiered_storage_s3_bucket_arn`
- `velero_s3_bucket_arn`

### Verifying this change

These changes are being verified against existing (upgrade/migration) and new environments.

### Documentation

Check the box below.

Need to update docs? 

- [X] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [X] `doc` 
  
  (If this PR contains doc changes)

### Todo
- [x] More testing around node security groups
- [x] Update README.md
- [ ] Write migration guide doc

### CODE REVIEW

File: tiered_storage.tf [hunk 1]
```diff
@@ -17,13 +17,28 @@
 # under the License.
 #
 
-terraform {
-  required_version = ">=1.0.0"
+resource "aws_s3_bucket" "tiered_storage" {
+  bucket = format("%s-offload", var.cluster_name)
```
Comment (gengmao) [inline tiered_storage.tf:21]: does this need an `snc-` prefix to comply vendor access policy?

File: velero.tf [hunk 2]
```diff
@@ -17,21 +17,31 @@
 # under the License.
 #
 
-data "aws_caller_identity" "current" {}
-
 resource "aws_s3_bucket" "velero" {
+  bucket = format("%s-cluster-backup", var.cluster_name)
```
Comment (gengmao) [inline velero.tf:21]: ditto

**Review by gengmao — APPROVED** [review]
LGTM.

File: tiered_storage.tf (hunk 1 above)
Comment (jrsdav) [reply tiered_storage.tf:21]: @gengmao that's a SNCloud specific configuration -- we control that using the `var.cluster_name` input, which will always include `*snc*` in it's name. Adding an additional prefix would be redundant.

**Comment by gengmao** [comment]
@jrsdav are `use_runtime_mode` and `create_iam_policies` default to `false` for new pool members? Keep them (or set them to `true`) just for pool members provisioned before this refactor, correct?

**Comment by jrsdav** [comment]
> @jrsdav are `use_runtime_mode` and `create_iam_policies` default to `false` for new pool members? Keep them (or set them to `true`) just for pool members provisioned before this refactor, correct?

@gengmao Good catch -- `use_runtime_policy` should be set to `false` since we want the parent EKS module to manage the cluster's IAM role for net-new clusters.

But the default behavior around `create_iam_policies` should be set to `true` to account for Platform/DIY users (otherwise the IAM roles will fail to provision correctly due to the IAM Policy `StreamNativeCloudRuntimePolicy` not being present).

**Comment by gengmao** [comment]
@jrsdav does the new version of community aws eks module still self-manages the amazon vpc cni plugin? Is it possible to switch to aws-managed k8s add-ons to avoid issues like https://github.com/streamnative/eng-support-tickets/issues/102 (part of https://github.com/streamnative/cloud-ops/issues/264)

**Comment by jrsdav** [comment]
> @jrsdav does the new version of community aws eks module still self-manages the amazon vpc cni plugin? Is it possible to switch to aws-managed k8s add-ons to avoid issues like [streamnative/eng-support-tickets#102](https://github.com/streamnative/eng-support-tickets/issues/102) (part of [streamnative/cloud-ops#264](https://github.com/streamnative/cloud-ops/issues/264))

@gengmao we will want to manage the plugin in self-managed mode as part of the bootstrap, as opposed to using the aws-managed plugin. This gives us more control over versions and allows for a cleaner adoption for existing self managed environments.

I have created an [issue ](https://github.com/streamnative/argo/issues/30) to track this as part of the ArgoCD work.

File: cluster_autoscaler.tf [hunk 3]
```diff
@@ -151,21 +159,21 @@ resource "helm_release" "cluster_autoscaler" {
       }
     ]
     image = {
-      tag = lookup(local.k8s_to_autoscaler_version, var.cluster_version, "v1.20.1") # image.tag defaults to the version corresponding to var.cluster_version's default value and must manually be updated
+      tag = lookup(local.k8s_to_autoscaler_version, var.cluster_version, "v1.21.1") # image.tag defaults to the version corresponding to var.cluster_version's default value and must manually be updated
```
Comment (jdmaguire) [inline cluster_autoscaler.tf:162]: I think we should default to 1.22 or 1.23.

File: variables.tf [hunk 4]
```diff
@@ -541,24 +418,36 @@ variable "node_termination_handler_chart_version" {
...
-  default     = 1
+  default     = 0
   description = "Desired number of worker nodes in the node pool."
   type        = number
 }
 
 variable "node_pool_disk_size" {
-  default     = 50
+  default     = 100
   description = "Disk size in GiB for worker nodes in the node pool. Defaults to 50."
```
Comment (jdmaguire) [inline variables.tf:451]:
```suggestion
  description = "Disk size in GiB for worker nodes in the node pool. Defaults to 100."
```

File: variables.tf [hunk 5]
```diff
@@ -138,7 +114,7 @@ variable "cluster_autoscaler_helm_chart_repository" {
 }
 
 variable "cluster_autoscaler_helm_chart_version" {
-  default     = "9.19.2"
+  default     = "9.21.0"
   description = "Helm chart version for the cluster-autoscaler. Defaults to \"9.10.4\". See https://github.com/kubernetes/autoscaler/tree/master/charts/cluster-autoscaler for more details."
```
Comment (jdmaguire) [inline variables.tf:118]: I think we should get rid of these `defaults to`. When we are generates the docs, terraform-docs actually does put the default in the table.

File: variables.tf [hunk 6]
```diff
@@ -208,75 +184,57 @@ variable "csi_settings" {
...
 }
 
-variable "enable_external_dns" {
-  default     = true
-  description = "Enables the External DNS addon service on the cluster. Defaults to \"true\", and in most situations is required by StreamNative Cloud."
+variable "disable_public_pulsar_endpoint" {
+  default     = false
+  description = "Whether or not to make the Istio Gateway use a public facing or internal network load balancer. If set to \"true\", additional configuration is required in order to manage the cluster from the StreamNative console"
   type        = bool
 }
```
Comment (jdmaguire) [inline variables.tf:215]: I like this ned variable name. `disable_ublic_pulsar_endpoint` is more description than the old name of `istio_network_loadbancer`

File: variables.tf [hunk 7]
```diff
@@ -316,107 +274,18 @@ variable "external_dns_settings" {
...
-  default     = 50
-  description = "Disk size in GiB for function worker nodes. Defaults to 20. Terraform will only perform drift detection if a configuration value is provided."
-  type        = number
-}
-
-variable "func_pool_disk_type" {
-  default     = "gp3"
-  description = "Disk type for function worker nodes. Defaults to gp3."
+variable "hosted_zone_id" {
+  default     = "*"
```
Comment (jdmaguire) [inline variables.tf:278]: Is `*` a good default?

File: cluster_autoscaler.tf [hunk 8]
```diff
@@ -109,11 +116,12 @@ locals {
     "1.20" = "v1.20.1",
     "1.21" = "v1.21.1",
     "1.22" = "v1.22.1",
+    "1.23" = "v1.23.0"
```
Comment (jdmaguire) [inline cluster_autoscaler.tf:119]:
```suggestion
    "1.23" = "v1.23.1"
```

The .1 is released. Should we use it?

Comment (jdmaguire) [reply cluster_autoscaler.tf:119]: 1.24.0 and 1.25.0 is also released. Should we add them to the map?

**Comment by addisonj** [comment]
I fixed the merge conflict

**Pull request merged**